### PR TITLE
[#2177] Migrate to Java 11 Syntax and Features

### PR DESCRIPTION
**Pull request by asdfghjkxd**
Part of #2177 

### Proposed commit message

```
Our current codebase is written on Java 8, which is around 2
generations behind on the latest supported LTS version for Java.

Let's move to migrate our CI/CD from Java 8 to Java 11 first to prepare
for any upcoming changes in our syntax, APIs and features.
```

### Other information

N/A

### CODE REVIEW

**Review by sopa301 — COMMENTED** [review]
I think we should also update the docs, particularly the required Java version in the developer guide!

**Comment by asdfghjkxd** [comment]
@sopa301 Done! I have updated the docs to reflect the migration from Java 8 to Java 11!

**Comment by asdfghjkxd** [comment]
@supermii2 I agree, we should discuss the different components that can be upgraded and begin implementing the changes in this PR!

**Review by sopa301 — APPROVED** [review]
LGTM!

**Comment by asdfghjkxd** [comment]
@supermii2 I have upgraded checkstyle to version `10.10.0` as described in #1304. Do let me know if there are any other areas that need upgrading!

**Comment by supermii2** [comment]
LGTM!

**Review by supermii2 — COMMENTED** [review]
LGTM!

File: build.gradle [hunk 1]
```diff
@@ -99,7 +99,7 @@ run {
 }
 
 checkstyle {
-    toolVersion = '9.3'
+    toolVersion = '10.10.0'
```
Comment (ckcherry23) [inline build.gradle:102]: I think it would be better if we do the checkstyle upgrade in a different PR, because there may be some changes that break backward compatibility.

Comment (asdfghjkxd) [reply build.gradle:102]: I see, sure thing, I will revert back to the previous commit before the introduction of the new checkstyle tool!

**Review by ckcherry23 — APPROVED** [review]
LGTM!

**Review by gok99 — COMMENTED** [review]
Really sorry for the review delays these last 2 weeks! Thanks for the thorough CI and docs updates. Looks most good to go!

Can we remove the `killDescendants` in `serveTestReportInBackground` at line 202 of `build.gradle`?

**Comment by asdfghjkxd** [comment]
@gok99 I have removed the flag from the `build.gradle` file. Do let me know if there are any more changes that needs to be made!

**Review by gok99 — APPROVED** [review]
LGTM

**Comment by ckcherry23** [comment]
Right now, the status checks are not passing because the branch protection rules require the "JDK 8" version to pass, whereas this PR changes the CI name to use "JDK 11". 

Hence, the "Branch protection rules" or "Status checks that are required" settings must be updated to reflect the new name for this PR to be merged.

**Comment by chan-j-d** [comment]
@ckcherry23 I've updated the branch protection rules to require the new JDK 11 checks instead. Though the frontend tests seem to be failing somewhat randomly, based on the history of CI runs on this PR. Should probably fix that soon and merge this as it breaks the other PRs.
